### PR TITLE
Combined fixes from other contributors, uses http api. Added PIR status, last alarm date, pic url, battery level for batterycamera and other fixes.

### PR DESCRIPTION
**Pull request by RenierM26**
Added fixes created by these contributors:
@FakeTruth and @Bjorntimmer and @balloob.

-Also added "BatteryCamera" category for C3A camera support. 
*Just note that polling will consume battery life.*

@BaQS, please review and commit to main branch? There are so many users that make use of this for Hassio. The work done by the other contributors make this addon compatible with Hassio again. 

Thanks for all the work you've done on this addon! I realize how much effort it is to reverse engineer APIs from apps due to lack of documentation.

### CODE REVIEW

**Comment by RenierM26** [comment]
Just some more info on the fixes in this pull request. (Again thanks to all the other contributors)

-Hikvision camera support
-Battery Camera type support
-Doesn't error out on null camera detection sensitivity returns.
-Privacy + Sleep IDs corrected and added in return. 
-uuid removed from requirements. (Not required on latest python versions - this is what breaks hassio integration)

**Review by RenierM26 — COMMENTED** [review]
1) Changed auth to general http login. (no need for fake agent as well as easier exploring of API via fidler on windows - just use IE as the site make use of JS which converts some json responses)
2) Added camera device info section. It returns raw json at the moment, need to add filtering via json loads function or eq)
3) Sometimes captcha blocks access to the site. (Just login via Internet Explorer on the https://ieu.ezvizlife.com site once and the script should work again)
4) I don't know what the "from" code in the header means but it looks fixed. Could be a MD5 hash of the url or fixed for a country. - Mabe something to explore.

**Review by RenierM26 — COMMENTED** [review]
-Rewrote most of camera.py to use less api calls.
-Added last alarm trigger and pic url!
-Disabled query device sensitivity for batterycamera (uses battery power)

File: pyezviz/__main__.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def main():
 
         # load camera object
         try:
-            client.login()
+            client._login()
             camera = EzvizCamera(client, args.serial)
```
Comment (FakeTruth) [inline pyezviz/__main__.py:128]: Missing positional argument camerainfo

File: requirements.txt [hunk 2]
```diff
@@ -1,2 +1,11 @@
-cryptography==2.8
-pandas==0.25.3
\ No newline at end of file
+certifi==2020.6.20
+chardet==3.0.4
+idna==2.10
+numpy==1.19.0
+pandas==1.0.5
+pyEzviz==0.1.5.7
```
Comment (FakeTruth) [inline requirements.txt:6]: Why is pyEzviz part of the requirements?

**Review by FakeTruth — COMMENTED** [review]
Nice work! Ran into some small issues when trying out your changes

File: pyezviz/__main__.py (hunk 1 above)
Comment (RenierM26) [reply pyezviz/__main__.py:128]: Hi There!

The EzvizCamera class needs to be called with an additional variable "camerainfo" 

EzvizCamera(self, device['deviceSerial'], camerainfo)

The json array "camerainfo" is created with the "_get_deviceinfo" function in the client.py file. I've done it this way as to reduce api calls. 

I've only tested with devices status. I see there are other calls to the EzvizCamera class, in the __Main__.py file that i missed . I'll work on fixing them.

I'm new to python so please excuse the minor errors. I have +- 2 weeks of experience so far but the basics seem simple enough.

File: requirements.txt (hunk 2 above)
Comment (RenierM26) [reply requirements.txt:6]: I'll update this file. Do you know if this file has any practical uses?

**Review by RenierM26 — COMMENTED** [review]
Fixed argument "camera" by fixing call to ezvizcamera class.

File: pyezviz/__main__.py (hunk 1 above)
Comment (RenierM26) [reply pyezviz/__main__.py]: Should work now.

File: requirements.txt (hunk 2 above)
Comment (RenierM26) [reply requirements.txt]: Removed pyEzviz from requirements file.

**Review by RenierM26 — COMMENTED** [review]
Fixed login check.

File: requirements.txt (hunk 2 above)
Comment (FakeTruth) [reply requirements.txt]: This file lets you install all dependencies for this project in one go:
```
pip3 install -r requirements.txt
```

File: pyezviz/__main__.py (hunk 1 above)
Comment (mindtripper) [reply pyezviz/__main__.py]: > Missing positional argument camerainfo

Following the instructions on using this with HA. I am seeing the same issue 
I believe the component has been loaded successfully with the correct version "Requirement already satisfied: **pyezviz** in /usr/local/lib/python3.8/site-packages **(0.1.5.9**)"

`2020-12-20 12:36:31 ERROR (MainThread) [homeassistant.components.camera] Error while setting up ezviz_my platform for camera
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 199, in _async_setup_platform
    await asyncio.shield(task)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/ezviz_my/camera.py", line 91, in setup_platform
    camera["ezviz_camera"] = EzvizCamera(ezviz_client, camera_serial)
TypeError: __init__() missing 1 required positional argument: 'camerainfo'`

Using the EZVIZ C6N camera.

Comment (mindtripper) [reply pyezviz/__main__.py]: Got it working by adding
`        cameras = ezviz_client.load_cameras()`

after line 51 in HA EZVIZ component (camera.py)
```
    try:
        ezviz_client = EzvizClient(account, password)
        ezviz_client.login()
        cameras = ezviz_client.load_cameras()
```

And changing former line 91 (now 92) from
`        camera["ezviz_camera"] = EzvizCamera(ezviz_client, camera_serial)`
to
`        camera["ezviz_camera"] = EzvizCamera(ezviz_client, camera_serial, camerainfo)`

Comment (FakeTruth) [reply pyezviz/__main__.py]: I think RenierM26 already fixed those issues in the latest revision. I only tested the command-line though, have not integrated it with Home Assistant.

Comment (mindtripper) [reply pyezviz/__main__.py]: Isn't 0.1.5.9 the latest version via master branch?
`pip install git+https://github.com/RenierM26/pyEzviz.git@master#pyezviz==0.1.5.6` is the version of the pyezviz I install.
(git is not my area of expertise).

Comment (FakeTruth) [reply pyezviz/__main__.py]: That's cool, I didn't know you could install packages from git with pip.
I tried this command in a clean environment and it installed pyEzviz 0.1.5.9 with RenierM26's fixes already applied:
```
pip install git+https://github.com/RenierM26/pyEzviz.git@master
```

Comment (mindtripper) [reply pyezviz/__main__.py]: Just found that out myself looking into this stuff. 
pyEzviz 0.1.5.9 in itself works just fine for me as well.

But Home Assistant EZVIZ component has the following line of code: `camera["ezviz_camera"] = EzvizCamera(ezviz_client, camera_serial)`. Which errors out. As RenierM26s pyEzviz 0.1.5.9 EzvizCamera require a third argument.
Ref:
```
class EzvizCamera(object):
    def __init__(self, client, serial, camerainfo):
        """Initialize the camera object."""
        self._client = client
        self._serial = serial
        self._camerainfo = camerainfo
```

Comment (FakeTruth) [reply pyezviz/__main__.py]: Oh I see, the Home Assistant component is missing something. Sorry I misunderstood the problem.
https://github.com/home-assistant/core/blob/dev/homeassistant/components/ezviz/camera.py#L91

**Comment by mindtripper** [comment]
Hi ReinierM26,

Great work on re-writing the component.
I've gotten it to partially work in Home Home Assistant 2020.12.1 by adding the missing argument in the HA EZVIZ component.
Initial status is correct and I can stream from the camera, but updates `def update(self):` are not fetched/presented in the HA gui.
Any thoughts on why that is?

**Comment by FakeTruth** [comment]
I suspect the original from BaQs also does not update properly.
The Home Assistant component periodically calls `status()` on `pyezviz.camera`, but if you look into `status()` you see this:
```
if not self._loaded:
    self.load()
```
`load()` is what downloads new data from the Ezviz API, but it's only called once, because it sets `_loaded` to 1 on success.

**Comment by mindtripper** [comment]
> I suspect the original from BaQs also does not update properly.
> The Home Assistant component periodically calls `status()` on `pyezviz.camera`, but if you look into `status()` you see this:
> 
> ```
> if not self._loaded:
>     self.load()
> ```
> 
> `load()` is what downloads new data from the Ezviz API, but it's only called once, because it sets `_loaded` to 1 on success.

But does this apply to the RenierM26 version of pyezviz? There is no `if not self._loaded` logic. Not that I can see.

Looking at the code at https://github.com/RenierM26/pyEzviz/blob/master/pyezviz/camera.py#L53
```
class EzvizCamera(object):
........
    def status(self):
        """Return the status of the camera."""
        self.load()

        return {
            'serial': self._serial,
........
```

Which is linked by the HA component
https://github.com/home-assistant/core/blob/dev/homeassistant/components/ezviz/camera.py#L131
```
class HassEzvizCamera(Camera):
.....
    def update(self):
        """Update the camera states."""

        data = self._ezviz_camera.status()
.....

```

Thank you for the input and discussion about this. Really appreciate it.

**Comment by FakeTruth** [comment]
Ah you're right. In this case, with RenierM26's changes the actual downloading of data is done by `EzvizClient._get_deviceinfo()`, see https://github.com/RenierM26/pyEzviz/blob/master/pyezviz/__main__.py#L126
This method is also called from `EzvizClient.load_cameras()` which you are using in the HA component.

What you could try to do is add these two lines to `EzvizCamera.status()` and add an extra parameter to force it to retrieve new data through the Ezviz API:
```
def status(self, update = False):
    """Return the status of the camera."""
    if update:
        self._camerainfo = self._client._get_deviceinfo()
        
    self.load()
    
    ....
```
Then in the HA component make the call like this: `data = self._ezviz_camera.status(True)`

It's not the prettiest fix, but it should do what you want for now.
Unfortunately there might be another problem due to the login session expiring after a while. So it might work for an hour, or days but eventually the session will expire and the API calls will fail until the script logs in again.

**Review by RenierM26 — COMMENTED** [review]
Added login checks

**Comment by mindtripper** [comment]
> self._camerainfo = self._client._get_deviceinfo()

I was just about to test the proposed workaround before I saw the recent changes by Renier26M. Using the pyEzviz-0.1.6.1build solves updates (stale data) and the need to add "camerainfo" argument in the HA component.

Again, big thanks to you @FakeTruth. I'm getting a better grasp of things based on your input and suggestions. 

@RenierM26 I don't need to change the HA component anymore to make EZVIZ work. This is great! I'll keep testing your updates and provide feedback.

**Comment by mindtripper** [comment]
Done tests with pyEzviz 0.1.6.1 and have found the following;

**--switch follow_move** does not work
pyezviz -u _redacted_ -p _redacted_ -d camera --serial _redacted_ -d switch --switch follow_move --enable 1
Partial debug result:
```
DEBUG:urllib3.connectionpool:https://ieu.ezvizlife.com:443 "POST /device/deviceSwitch!configDeviceSwitchStatus.action HTTP/1.1" 200 19
('Could not set the switch, maybe a permission issue ?: Got %s : %s)', '200', '{"resultCode":"-1"}')

```
**alarm --notify 1** does not work
Partial debug result:
```
send: b'clientType=1&infoDetail=%7B%22operationType%22%3A1%2C%22detail_redacted_deviceSerial_redacted_&infoType=3&netType=WIFI&requestType=0'
reply: 'HTTP/1.1 404 Not Found\r\n'

send: b'subSerial=<_redacted_>&type=0&value=3'
reply: 'HTTP/1.1 400 Bad Request\r\n'

DEBUG:urllib3.connectionpool:https://ieu.ezvizlife.com:443 "POST /api/device/configAlgorithm HTTP/1.1" 400 456

```

**alarm --sound 0** and **alarm --sensibility** behave as follows;
```
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sound 0
"alarm_sound_mod": "MUTE_MODE",
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sound 1
"alarm_sound_mod": "ALARM_MODE",
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sound 2
"alarm_sound_mod": "MUTE_MODE",
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 0
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 1
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 2
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 3
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 4
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 5
pyezviz -u _redacted_ -p _redacted_ camera --serial _redacted_ alarm --sensibility 6
"detection_sensibility": null,

```

**Comment by RenierM26** [comment]
Hi @mindtripper,

Could you please share the model number of your cameras? 

The alarm functions (Alarm, detection sensitivity and sound)   should work with v0.1.6.2

**Comment by mindtripper** [comment]
> Hi @mindtripper,
> 
> Could you please share the model number of your cameras?
> 
> The alarm functions (Alarm, detection sensitivity and sound) should work with v0.1.6.2

I’m indisposed for a couple of days, but I will test as soon as I can.

Model: EZVIZ C6N
ASIN: B07W4FMSD9

**Comment by RenierM26** [comment]
Hi All,

I've created a HACS compatible Hassio extension that makes use of my fork of pyezviz. 

*I've added services for all the items that can be set via pyezviz with proper examples.
*Also added the extra attributes that are now available via pyezviz.
*Might have also fixed the 12 hour "cookie expiry" problem with the addon. *Busy testing this fix*

Just add this custom HACS repository if you'll like to test:
RenierM26/ha-ezviz

The code still needs some optimization. I'm considering adding my own version to pypi and adding a pull request to the official HA repository. Don't want to step on anybodies toes but it seems like BaQs might have lost interest in this project -- and there are a lot of people really interested in getting the ezviz component working again.

Still needs a lot of work but at least it's mostly working at the moment. I'm learning async python functions and all the various ha specific classes/variables available.

**Comment by RenierM26** [comment]
Hi All,

I have created a new pypi project with corresponding github repository. 

API:
Github: https://github.com/RenierM26/py-Ezviz
pypi project: https://pypi.org/project/py-ezviz/

Hassio Component page:
Github: https://github.com/RenierM26/ha-ezviz

**Comment by mindtripper** [comment]
> Hi @mindtripper,
> 
> Could you please share the model number of your cameras?
> 
> The alarm functions (Alarm, detection sensitivity and sound) should work with v0.1.6.2

@RenierM26 did a test using v0.1.6.4 seeing the same results. Quick summary below.

```
alarm --sensibility "0-9" executes successfully, but status indicates that nothing has been set.

{
  "serial": "...",
  "name": "C6N(...)",
  "version": "V5.3.0 build 200514",
  "upgrade_available": 0,
  "device_sub_category": "C6N",
  "detection_sensibility": null,
}
```

```
switch --switch follow_move --enable "0,1"
('Could not set the switch, maybe a permission issue ?: Got %s : %s)', '200', '{"resultCode":"-1"}')

```

`alarm --sound "0-2"` works.
`alarm --notify "0,1" and ha-ezviz (camera.disable_motion_detection and camera.enable_motion_detection)` works.

**Comment by BaQs** [comment]
Woh!! what an impressive request!

I've been very busy recently. Would you mind rebasing ?
I also added you to the official contributors.

Best!

**Comment by RenierM26** [comment]
Latest changes where merged by BaQs. Cleaning up this stale push request.

**Pull request closed without merging**